### PR TITLE
Loader: Add classes option

### PR DESCRIPTION
**Pull request by gabrielschulhof**
Fixes gh-7699

### CODE REVIEW

File: js/widgets/loader.js [hunk 1]
```diff
@@ -53,14 +56,14 @@
 			text: "loading"
 		},
 
-		defaultHtml: "<div class='" + loaderClass + "'>" +
+		defaultHtml: "<div>" +
 			"<span class='ui-icon-loading'></span>" +
 			"<h1></h1>" +
```
Comment (arschmitz) [inline js/widgets/loader.js]: Any elements we create need a class name

File: js/widgets/loader.js [hunk 2]
```diff
@@ -53,14 +56,14 @@
...
 			"</div>",
 
 		// For non-fixed supportin browsers. Position at y center (if scrollTop supported), above the activeBtn (if defined), or just 100px from top
 		fakeFixLoader: function() {
-			var activeBtn = $( "." + $.mobile.activeBtnClass ).first();
+			var activeBtn = $( ".ui-button-active" ).first();
```
Comment (arschmitz) [inline js/widgets/loader.js]: we don't support any browsers that dont support fixed position anymore

File: js/widgets/loader.js [hunk 3]
```diff
@@ -77,7 +80,7 @@
 				screenHeight = $.mobile.getScreenHeight();
 
 			if ( offset.top < scrollTop || ( offset.top - scrollTop ) > screenHeight ) {
```
Comment (arschmitz) [inline js/widgets/loader.js]: We no longer support browser without fixed position

File: js/widgets/loader.js [hunk 4]
```diff
@@ -154,10 +158,10 @@
 		},
 
 		hide: function() {
-			$html.removeClass( "ui-loading" );
+			this._removeClass( html, "ui-loading" );
 
 			if ( this.options.text ) {
-				this.element.removeClass( "ui-loader-fakefix" );
+				this._removeClass( "ui-loader-fakefix" );
```
Comment (arschmitz) [inline js/widgets/loader.js]: no longer support browsers without fixed position support

File: tests/unit/loader/loader_core.js [hunk 5]
```diff
@@ -41,11 +41,12 @@ $.testHelper.delayStart();
...
 
-	test( "showing the loader does not hide the text when the loading message is true", function(){
-		$.mobile.loading( "option", "textVisible", true ).loader( "show" );
+	test( "showing the loader does not hide the text when the loading message is true",
+		function( assert ) {
+			$.mobile.loading( "option", "textVisible", true ).loader( "show" );
 
-		ok($(".ui-loader").hasClass( "ui-loader-verbose" ), "displaying text");
-	});
+			assert.hasClasses( $( ".ui-loader" )[ 0 ], "ui-loader-verbose", "displaying text" );
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: don't pass the native element for no reason

File: tests/unit/loader/loader_core.js [hunk 6]
```diff
@@ -108,31 +111,50 @@ $.testHelper.delayStart();
...
 			.loading( "option", "text", "fozzle" )
 			.loader( "option", "theme", "x" )
 			.loader( "show" );
 
-		ok($(".ui-loader").hasClass( "ui-body-x" ), "has theme x");
+		assert.hasClasses( $( ".ui-loader" )[ 0 ], "ui-body-x", "has theme x" );
 		deepEqual($(".ui-loader h1").text(), "fozzle", "has text fozzle in loading config object");
 	});
+
+	test( "Loader sets itself as fakefix when it's off-center", function( assert ) {
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: no longer support browsers without position fixed

File: tests/unit/loader/loader_core.js [hunk 7]
```diff
@@ -108,31 +111,50 @@ $.testHelper.delayStart();
...
+		$( ".ui-loader" )
+			.css({ left: -10000, top: -10000 });
+
+		$.mobile.loading( "checkLoaderPosition" );
+
+		assert.hasClasses( $( ".ui-loader" )[ 0 ], "ui-loader-fakefix",
+			"loader has class ui-loader-fakefix" );
+
+		$( ".ui-loader" ).removeAttr( "style" );
+	});
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: missing space

File: tests/unit/loader/loader_core.js [hunk 8]
```diff
@@ -108,31 +111,50 @@ $.testHelper.delayStart();
...
-		ok($(".ui-loader").hasClass( "ui-body-x" ), "has theme x");
+		assert.hasClasses( $( ".ui-loader" )[ 0 ], "ui-body-x", "has theme x" );
 		deepEqual($(".ui-loader h1").text(), "fozzle", "has text fozzle in loading config object");
 	});
+
+	test( "Loader sets itself as fakefix when it's off-center", function( assert ) {
+		$.mobile.loading( "show" );
+
+		$( ".ui-loader" )
+			.css({ left: -10000, top: -10000 });
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: missing spaces

**Comment by arschmitz** [comment]
Most of the same issues here as in the page widget classes PR

**Comment by arschmitz** [comment]
@gabrielschulhof rebasing with master blew this up can you fix it and ill review again?

File: js/widgets/loader.backcompat.js [hunk 9]
```diff
@@ -0,0 +1,64 @@
...
+			"</div>",
+
+		show: function( theme ) {
+			var html;
+
+			this._superApply( arguments );
+
+			html = ( $.type( theme ) === "object" && theme.html || this.options.html );
+
+			// TODO verify that jquery.fn.html is ok to use in both cases here
```
Comment (arschmitz) [inline js/widgets/loader.backcompat.js]: I would not say this is really a todo any more since its not in backcompat

File: js/widgets/loader.js [hunk 10]
```diff
@@ -41,54 +39,52 @@ return $.widget( "mobile.loader", {
...
 	options: {
+		classes: {
+			"ui-loader": "ui-corner-all",
+			"ui-loader-icon": "ui-icon-loading"
+		},
+
+		enhanced: false,
+
 		// the theme for the loading message
 		theme: "a",
```
Comment (arschmitz) [inline js/widgets/loader.js]: use theme extension

File: js/widgets/loader.js [hunk 11]
```diff
@@ -41,54 +39,52 @@ return $.widget( "mobile.loader", {
...
-	resetHtml: function() {
-		this.element.html( $( this.defaultHtml ).html() );
-	},
+	// DEPRECATED as of 1.5.0 and will be removed in 1.6.0 - we no longer support browsers
+	// incapable of native fixed support
+	fakeFixLoader: $.noop,
+
+	// DEPRECATED as of 1.5.0 and will be removed in 1.6.0 - we no longer support browsers
+	// incapable of native fixed support
+	checkLoaderPosition: $.noop,
```
Comment (arschmitz) [inline js/widgets/loader.js]: these should be in the backcompat extension

File: js/widgets/loader.js [hunk 12]
```diff
@@ -119,50 +113,30 @@ return $.widget( "mobile.loader", {
...
-		if ( loadSettings.html ) {
-			this.element.html( loadSettings.html );
-		} else {
-			this.element.find( "h1" ).text( message );
-		}
+		this._removeClass( "ui-loader-verbose ui-loader-default ui-loader-textonly" )
+			._addClass( "ui-loader-" +
+			( textVisible || msgText || theme.text ? "verbose" : "default" ) +
+			( loadSettings.textonly || textonly ? " ui-loader-textonly" : "" ),
+				"ui-body-" + theme );
```
Comment (arschmitz) [inline js/widgets/loader.js]: use theme extension

File: js/widgets/loader.js [hunk 13]
```diff
@@ -119,50 +113,30 @@ return $.widget( "mobile.loader", {
...
+
+		this._ui.header.text( message );
 
 		// If the pagecontainer widget has been defined we may use the :mobile-pagecontainer
 		// and attach to the element on which the pagecontainer widget has been defined. If not,
 		// we attach to the body.
 		this.element.appendTo( $.mobile.pagecontainer ?
 			$( ":mobile-pagecontainer" ) : $( "body" ) );
```
Comment (arschmitz) [inline js/widgets/loader.js]: Add todo to update this once https://github.com/jquery/jquery-mobile/pull/7947 lands

File: tests/unit/loader/loader_core.js [hunk 14]
```diff
@@ -30,9 +34,10 @@ $( document ).one( "mobileinit", function() {
 } );
 
 test( "prototype options are used for mobile loader", function() {
-	$.mobile.loading( 'show' );
+	$.mobile.loading( "show" );
 
-	deepEqual( $( '.ui-loader' ).text(), "mobileinit", "prototype options set the text and make it visible" );
+	deepEqual( $( ".ui-loader" ).text(), "mobileinit",
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: no need for deepEqual

File: tests/unit/loader/loader_core.js [hunk 15]
```diff
@@ -41,17 +46,19 @@ test( "showing the loader does not show the text when the loading message is fal
...
 
-test( "hiding the loader doesn't add the dialog to the page when loading message is false", function() {
-	$.mobile.loading( "option", "text", true ).loader( "show" ).loader( "hide" );
+test( "hiding the loader doesn't add the dialog to the page when loading message is false",
+	function() {
+		$.mobile.loading( "option", "text", true ).loader( "show" ).loader( "hide" );
 
-	deepEqual( $( ".ui-loading" ).length, 0, "page should not be in the loading state" );
-} );
+		deepEqual( $( ".ui-loading" ).length, 0, "page should not be in the loading state" );
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: no need for deepEqual

File: tests/unit/loader/loader_core.js [hunk 16]
```diff
@@ -76,10 +83,12 @@ test( "page loading should contain custom loading message when used in param obj
...
 
-test( "page loading should contain different theme when used in param object", function() {
-	$.mobile.loading( "show", { theme: "l" } );
-	ok( $( ".ui-loader" ).hasClass( "ui-body-l" ), "loader has theme l" );
-} );
+test( "page loading should contain different theme when used in param object",
+	function( assert ) {
+		$.mobile.loading( "show", { theme: "l" } );
+
+		assert.hasClasses( $( ".ui-loader" ), "ui-body-l", "loader has theme l" );
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: message seems unneeded use default

File: tests/unit/loader/loader_core.js [hunk 17]
```diff
@@ -88,10 +97,12 @@ test( "page loading should contain new html when provided, prefers passed param"
 } );
 
 test( "page loading should always contain text when passed as the second arg", function() {
+
 	// simulate error call in navigation ajax error callback
 	$.mobile.loading( "option", "textVisible", true ).loader( "show", "e", "foo serious", true );
 
-	deepEqual( $( ".ui-loader" ).text(), "foo serious", "loader has message regardless of global setting" );
+	deepEqual( $( ".ui-loader" ).text(), "foo serious",
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: no need for deepEqual

File: tests/unit/loader/loader_core.js [hunk 18]
```diff
@@ -105,34 +116,41 @@ test( "page loading should always contain text when passed as an object prop", f
 			textonly: true
 		} );
 
-	deepEqual( $( ".ui-loader" ).text(), "foo serious second", "loader has message regardless of global setting" );
+	deepEqual( $( ".ui-loader" ).text(), "foo serious second",
```
Comment (arschmitz) [inline tests/unit/loader/loader_core.js]: no need for deepEqual

File: js/widgets/loader.js (hunk 12 above)
Comment (gabrielschulhof) [inline js/widgets/loader.js]: This function takes the swatch to be shown as an argument. The theme extension only handles ```_create()``` and ```_setOptions()```.

**Comment by arschmitz** [comment]
@gabrielschulhof i don't see any separate backcompat tests

**Comment by gabrielschulhof** [comment]
@arschmitz, you're right - I left them out.

**Comment by arschmitz** [comment]
:+1:

**Pull request closed without merging**